### PR TITLE
updated test_directed_edge_swap #5814 

### LINKED ISSUE #5814: Improve testing for directed_edge_swap
There is an opportunity to improve the testing of the `directed_swap_function`.

There is a concrete suggestions [here](https://github.com/networkx/networkx/pull/5663#discussion_r906225024) for improving `test_directed_edge_swap` by validating that the input and output edges are indeed different. Upon reflection, I think this check could even be improved further by verifying that the number of swaps is correct... I think the difference in the edge sets should be 2*`nswap` - that might be one way to check.

It'd also be good to add more test cases of non-path graphs where there are only a limited number of edge swaps possible, and verify that the algorithm can recover all of them. See [the discussion](https://github.com/networkx/networkx/pull/5663#pullrequestreview-1018715617) for further details.

### PR DESCRIPTION
**Pull request by Dishie2498**
Please review

fixes #5814

### CODE REVIEW

**Review by rossbar — COMMENTED** [review]
IMO this is far too complicated for a unit test. It's not clear what this is intended to test; also why remove the original test? Removing tests like this is always a difficult proposition because it may be covering paths that you don't immediately expect.

**Comment by Dishie2498** [comment]
The original test has not been removed (The only difference is that `path_graph` is not used to create the path graph). The test is still being run on a path graph with 200 nodes. The only addition to the test for the path graph is that, the input and output edges are being compared for differences.
Also, the size of the function seems large because, `directed_edge_swap` will not return a graph for every complete DiGraph given as input. Hence, the code: 
Firstly creates a complete DiGraph (with 200 nodes) in which edge swapping is actually possible.
Secondly, calls the `directed_edge_swap` on it
Lastly, confirms whether the returned graph is correct or not.

**Comment by rossbar** [comment]
I'm still -1 on adding this - this makes the unit test much harder to follow.

**Comment by dschult** [comment]
Is this the same as:
```python
    PG = nx.path_graph(200, create_using=nx.DiGraph)
    # CG is a DiGraph with one edge direction of each pair of nodes
    CG = nx.DiGraph(nx.complete_graph(200).edges)
    for G in (PG, CG):
        H = G.copy()
        in_degrees = sorted(H.in_degree())
        out_degrees = sorted(H.out_degree())
        nx.directed_edge_swap(G, nswap=40, max_tries=500, seed=1)
        assert in_degrees == sorted(G.in_degree())
        assert out_degrees == sorted(G.out_degree())
```

This approach only checks the degree sequences and not whether the right number of edges were swapped.
If we want to get the right number of edges swaps, you might be able to get a list of them using:
`G.edges - H.edges` and `H.edges - G.edges`. 

But I'm not sure we can check for the right number of swaps... I believe it is possible for the algorithm to randomly select a swap that un-does what a previous swap did.

**Comment by Dishie2498** [comment]
The problem is that edge swaps in a directed complete graph (ie. where each edge is bidirectional and there are pairs of parallel edges) are not possible as there will be no difference in the input and output graphs. Hence, the `directed_edge_swap()` function will technically fail in such a case. 

Earlier, I chose to test a graph which is complete, in the sense of connections between every vertex but there are no parallel edges so that, swapping is possible. This was built using my custom code as I couldn't find a way to do this using any inbuilt functions.
Please let me know if there are other types of directed graphs in which swapping would be allowed and would be appropriate to test with this function. As an exercise I've tried `balanced_tree` and `binomial_tree` and they seem to be working fine.

**Comment by Dishie2498** [comment]
I tried the to test the following graphs:
Cyclic graph
Random_tree
Path graph
Havel hakimi graph 
Balanced tree
Binomial tree

Like you suggested, I used small graphs. Finally, I checked for 3 things:
-> in-degree consistency
-> out-degree consistency
-> change in edge != NULL

I've pushed the code, please let me know what improvements can be made.

**Comment by Dishie2498** [comment]
Also, I initially used the 'edges equal' function. However it regards 0->1 and 1->0 as equal.
When comparing edges in a directed graph, this results in a false positive. 
Do you think we should add another function to be able to check for edge equality in a directed graph?

**Comment by dschult** [comment]
For these tests, I think the more obvious check using `G.edges == H.edges` should work.

I'm not sure how much the `nx.utils.edges_equal` is used to compare two graphs (where the above syntax works better). That function is usually used for comparing a list of edges with `G.edges`.  At least I think that is true. Note that for a set of edges, `G.edges == set_of_edges` also works well. But I bet there are a lot of places where this utility function is used instead.  For now, I would suggest just using the direct equals operator and we can talk about updating the utility script.

**Comment by Dishie2498** [comment]
Thanks for the feedback! I have changed the assert.

**Comment by Dishie2498** [comment]
@dschult @rossbar 
Could you please review the changes?

Also, about the utility function, `edges_equal` considering a->b and b->a as different edges, should I create an issue regarding that and start working on it?

File: networkx/algorithms/tests/test_swap.py [hunk 1]
```diff
@@ -4,12 +4,20 @@
...
+    binomial = nx.binomial_tree(3, create_using=nx.DiGraph)
+    graph = nx.directed_havel_hakimi_graph(
+        [1, 2, 1, 2, 2, 2], [3, 1, 0, 1, 2, 3], create_using=None
+    )
+    balanced_tree = nx.balanced_tree(2, 3, create_using=nx.DiGraph)
+
+    for G in (path, binomial, graph, cycle, tree, balanced_tree):
+        swapped = nx.directed_edge_swap(G.copy())
+        assert sorted(G.in_degree()) == sorted(swapped.in_degree())
+        assert sorted(G.in_degree()) == sorted(swapped.in_degree())
```
Comment (MridulS) [inline networkx/algorithms/tests/test_swap.py:19]: Shouldn't one be out_degrees?

**Review by dschult — APPROVED** [review]
I pushed a commit that will hopefully put this PR over the hump.
I rewrote the test so it checks both in and out degree and checks that 3 edges have changed for 1 edge_swap. I also added a test that checks that a second edge_swap can undo the first edge swap. And I added a line to the doc_string that tells the user later swaps are allowed to undo previous swaps. 

I think this is ready.

**Review by rossbar — APPROVED** [review]
LGTM, thanks @Dishie2498 and @dschult !

**Pull request merged**